src/ImageSharp/ColorSpaces/Conversion/Implementation/Icc/Calculators/ClutCalculator.cs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using System.Numerics;
+using SixLabors.ImageSharp.Metadata.Profiles.Icc;
+
+namespace SixLabors.ImageSharp.ColorSpaces.Conversion.Icc;
+
+internal class ClutCalculator : IVector4Calculator
+{
+    private readonly int inputCount;
+    private readonly int outputCount;
+    private readonly float[][] lut;
+    private readonly byte[] gridPointCount;
+    private readonly int[] indexFactor;
+    private readonly int nodeCount;
+    private readonly float[][] nodes;
+
+    public ClutCalculator(IccClut clut)
+    {
+        Guard.NotNull(clut, nameof(clut));
+
+        this.inputCount = clut.InputChannelCount;
+        this.outputCount = clut.OutputChannelCount;
+        this.lut = clut.Values;
+        this.gridPointCount = clut.GridPointCount;
+        this.indexFactor = CalculateIndexFactor(clut.InputChannelCount, clut.GridPointCount);
+        this.nodeCount = (int)Math.Pow(2, clut.InputChannelCount);
+        this.nodes = new float[this.nodeCount][];
+    }
+
+    public unsafe Vector4 Calculate(Vector4 value)
+    {
+        Vector4 result = default;
+        this.Interpolate((float*)&value, this.inputCount, (float*)&result, this.outputCount);
+
+        return result;
+    }
+
+    private static int[] CalculateIndexFactor(int inputCount, byte[] gridPointCount)
+    {
+        int[] factors = new int[inputCount];
+        int gpc = 1;
+        for (int j = inputCount - 1; j >= 0; j--)
+        {
+            factors[j] = gpc * (gridPointCount[j] - 1);
+            gpc *= gridPointCount[j];
+        }
+
+        return factors;
+    }
+
+    private unsafe void Interpolate(float* values, int valueLength, float* result, int resultLength)
+    {
+        for (int i = 0; i < this.nodes.Length; i++)
+        {
+            int index = 0;
+            for (int j = 0; j < valueLength; j++)
+            {
+                float fraction = 1f / (this.gridPointCount[j] - 1);
+                int position = (int)(values[j] / fraction) + ((i >> j) & 1);
+                index += (int)((this.indexFactor[j] * (position * fraction)) + 0.5f);
```

Comment:
Just gave this a quick look and haven't fully digested how it's supposed to work yet, but this `+ 0.5f` looks wrong.  You'll be looking for a floor value to start and then finding the point after it to interpolate your fractional value.

Reply:
>Are you clear on the steps involved for CMYK->RGB conversion, and have you checked that your inputs to the CLUT step are correct?

I thought the process was the same regardless of input. Identify the correct calculator, convert to a profile connection space, then from that to the target profile? 

Are the inputs correct? I'm not 100% sure. There's a method in the calculator `CalculateIndexFactor` that generates the `indexFactor` array. The number in the first position there is so high that the calculated value is over 1000 greater than the length of the `lut` array. I can't find the equivalent in DemoIccMAX to that method. 

I'll have a look at those Little-CMS sources. Brian will be much more capable than I am reading them though.

Reply:
I think something might be wrong with the values we've parse for the gridpointcount array. They appear to be all set to 9 but the ICCProfile inspector tool found here https://www.color.org/profileinspector.xalter maxes out at 8 for each channel.

Reply:
The input values seem correct. I compared them to the values read in https://www.color.org/profileinspector.xalter The issue appears to be in the calculator itself.